### ISSUE TITLE
Documentation on 1d integration over an array

### ISSUE THREAD

**Issue by ACharletENSL** [nschloe/quadpy#381]
quadpy offers the possibility to integrate a function returning an array, which is one of the simplest case and the reason that made me give quadpy a shot (same case as described in this stackoverflow post https://stackoverflow.com/questions/34346809/integrating-functions-that-return-an-array-in-python )
But there is no documentation nor examples explaining clearly how to do this, it would help greatly as it is a very common problem

**Comment by nschloe** [comment]
What have you tried?

**Comment by ACharletENSL** [comment]
With integrand(x) a function taking a float and returning a 3d array, I tried `integral = quadpy.quad(integrand, 1.,np.infty)[0]` among other things and was encountering broadcasting problems during integration:
ValueError: operands could not be broadcast together with shapes (63,101,101) (21,)

But while writing this comment I tried
```
scheme = quadpy.line_segment.gauss_laguerre(3)
integral = scheme.integrate(integrand, [1., np.infty])
```
which just worked perfectly. As I thought it was really easy but I lacked a few lines explaining this basic method in the documentation

**Comment by nschloe** [comment]
Looks like you're using a mega outdated version of quadpy.

**Comment by ACharletENSL** [comment]
I'm using v0.16.1

**Comment by nschloe** [comment]
`line_segment` is not in 0.16. What's  your output of
```
print(quadpy.__version__)
```
?

**Comment by ACharletENSL** [comment]
oops i wrote line_segment based on an old doc I think but in the code it's c1, I forgot to correct this in my comment
Sorry for the confusion

**Comment by nschloe** [comment]
Well where's the MWE then??

**Comment by ACharletENSL** [comment]
Well, I was tinkering a bit with my functions to make them integrable by quadpy (as I have severals and only managed to integrate the simplest one without seeing a 'ValueError: operands could not be broadcast together'), but I broke the case that was functioning while vectorizing my basic functions and am now getting AssertionError for all of them
So this is how my code is written, where integ is a numpy array of shape (Nx,Ny,Nz), params is a stack of 9 such arrays and x is a float:
```
scheme = quadpy.c1.gauss_legendre(3)

def function(params):
  def integrand(x,params):
    return integ
  res = scheme.integrate(integrand, [1.,np.infty])
  return res
```
with a lot of operations involving the variable x in the definition of the integrand. I managed to modify my integrand function so she may return res of shape (p, Nx, Ny, Nz) when given an array of shape (p,) instead of the float x, but that won't work

**Comment by nschloe** [comment]
The MWE is incomplete. I get
```
NameError: name 'quadpy' is not defined
```

**Comment by ACharletENSL** [comment]
I added the importation line in my example

**Comment by nschloe** [comment]
I run the code and find no error.

**Comment by ACharletENSL** [comment]
This isn't working, i'm still getting the following error:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/acharlet/.local/lib/python3.8/site-packages/quadpy/c1/_helpers.py", line 68, in integrate
    domain_shape, range_shape, interval_set_shape = _find_shapes(
  File "/home/acharlet/.local/lib/python3.8/site-packages/quadpy/c1/_helpers.py", line 46, in _find_shapes
    assert fx.shape == range_shape + interval_set_shape + x.shape
AssertionError
```

Here are the lines of code, the variables Density to VZ are 3d arrays, Frequency and x are floats, integrand(x) is a 3d array with same shape as Density and the others
```
def integrand(x):
    	return rF.IC_1eRad(Density,Temperature,X_Coord,Y_Coord,Z_Coord,VX,VY,VZ,Frequency,x)

Integral = scheme.integrate(integrand, [1.,np.infty])
```

**Comment by nschloe** [comment]
The code you posted [here](https://github.com/nschloe/quadpy/issues/381#issuecomment-685467659) does _nothing_. It just imports quadpy, initializes a scheme and defines a function. From this, you cannot get the error you describe. You have run something else. Unfortunately I have no time for guessing what you have actually run.

**Comment by ACharletENSL** [comment]
Of course it does nothing
The question I'm really asking is how to use quadpy to integrate a function f(x,float,array1,..array9) returning an array along the float variable x. How should it be written and what are the underlying contions for f that are needed for the integration scheme to work ?

**Comment by nschloe** [comment]
https://github.com/nschloe/quadpy/issues/381#issuecomment-684602166

**Comment by ACharletENSL** [comment]
I'm not running anything else except numpy operations on my variables, and use `scheme.integrate(integrand, [1.,np.infty])` as written two comments before.

The following code can be run everywhere and is the exact same algorithm i'm using, I only changed the integration limits to have a real value in the end (with the following lines my integrand goes to infinity when x grows, where in my code it diminishes quickly to zero) but am still getting the same AssertionError.
For clarification: 
- ElectronDistribution takes two (Nx,Ny,Nz) arrays and a float (Lorentz), returns an (Nx,Ny,Nz) array. If you input a (p) vector instead of a float you'll get a (p,Nx,Ny,Nz) array
- IC_1eRad takes two floats (Frequency, Lorentz) and 9 (Nx,Ny,Nz) arrays, returns an (Nx,Ny,Nz) array. If you input a (p) vector for the variable Lorentz you'll get a (p,Nx,Ny,Nz) array

```
import numpy as np
import quadpy
scheme = quadpy.c1.gauss_legendre(3)
Nx,Ny,Nz = 63, 101, 101
Density = np.ones((Nx,Ny,Nz))
Temperature = np.ones((Nx,Ny,Nz))
X_Coord = np.ones((Nx,Ny,Nz))
Y_Coord = np.ones((Nx,Ny,Nz))
Z_Coord = np.ones((Nx,Ny,Nz))
VX = np.ones((Nx,Ny,Nz))
VY = np.ones((Nx,Ny,Nz))
VZ = np.ones((Nx,Ny,Nz))
Frequency = 1.
PlanckConst = 1.
BoltzmannConst = 1.
TemperatureRadiationSource = 1.
ElectronMass = 1.
SigmaThomson = 1.
def derive_NElectron(Density):
    return Density

def derive_InvNormalizedTemperature(Temperature):
    return Temperature

def EoverK2(InvNormalizedTemperature):
    return InvNormalizedTemperature

def derive_ICIncidentAngle(X_Coord,Y_Coord,Z_Coord,VX,VY,VZ):
    return np.full(X_Coord.shape, np.pi/3.)

def ElectronDistribution(Density,Temperature,Lorentz):
    Lorentz = np.array(Lorentz)
    NElectron = derive_NElectron(Density) 
    InvNormalizedTemperature=derive_InvNormalizedTemperature(Temperature)
    Ones = np.ones(Density.shape)
    vecLorentz = np.multiply.outer(Lorentz, Ones)
    Lorentz2 = np.multiply.outer(Lorentz**2, Ones)
    expLorentz = np.multiply.outer(np.exp(Lorentz), Ones)
    return NElectron*InvNormalizedTemperature*vecLorentz*np.sqrt(Lorentz2-1.)*expLorentz*EoverK2(InvNormalizedTemperature)

def IC_1eRad(Density,Temperature,X_Coord,Y_Coord,Z_Coord,VX,VY,VZ,Frequency,Lorentz):
    '''Radiated inverse Compton scattering power from one electron of given frequency and Lorentz'''
    Lorentz = np.array(Lorentz)
    IncAngle = derive_ICIncidentAngle(X_Coord,Y_Coord,Z_Coord,VX,VY,VZ)
    Beta = np.sqrt(1. - Lorentz**-2)
    Geom = 1. - np.multiply.outer(Beta,np.cos(IncAngle))
    Vec = np.ones(Lorentz.shape)
    Lorentz2 = np.multiply.outer(Lorentz**2,np.ones(Density.shape))
    Lorentz4 = np.multiply.outer(Lorentz**4,np.ones(Density.shape))
    Fac = 4. * ElectronMass * SigmaThomson / PlanckConst
    InitFreq = Frequency/(Lorentz2*Geom)
    Planck=1./(np.exp(PlanckConst * InitFreq / (BoltzmannConst * TemperatureRadiationSource)))
    EDistribution = ElectronDistribution(Density,Temperature,Lorentz)
    return Fac*Planck*Geom*EDistribution/(Frequency*np.sqrt(np.multiply.outer(Vec,np.cos(IncAngle)**4)+8.*Lorentz4*Geom))

def IC_Emission(Density,Temperature,X_Coord,Y_Coord,Z_Coord,VX,VY,VZ,Frequency):
    '''Volumic power emission per unit frequency'''
    def integrand(x):
    	return IC_1eRad(Density,Temperature,X_Coord,Y_Coord,Z_Coord,VX,VY,VZ,Frequency,x)
    Integral = scheme.integrate(integrand, [1.,10.])
    return Integral

res = IC_Emission(Density,Temperature,X_Coord,Y_Coord,Z_Coord,VX,VY,VZ,Frequency)
```

**Comment by nschloe** [comment]
This example is complete, but it's not minimal. Nobody will want to dig through your functions and variables. I'm sure your can shorten this to only a few lines. Also a tip: Use [black](https://github.com/psf/black), this increases readability, and stick to Python conventions of only using lowercase + underscore for variable and function names.

**Comment by ACharletENSL** [comment]
Sorry, I'm not an informatician so I may not have all the good reflexes. Here is a smaller working example that returns me the same error:

```
import numpy as np
import quadpy
scheme = quadpy.c1.gauss_legendre(3)
nx,ny,nz = 63, 101, 101
array1 = np.ones((nx,ny,nz))
array2 = np.ones((nx,ny,nz))
array3 = np.ones((nx,ny,nz))
param = 1.

def func1(array1,array2,x):
	x = np.array(x)
	vecX = np.multiply.outer(x, np.ones(array1.shape))
	return array1*array2*vecX

def func2(array1,array2,array3,param,x):
	x = np.array(x)
	idX = np.multiply.outer(x, np.ones(array1.shape))
	expX = np.multiply.outer(np.exp(-x), np.ones(array1.shape))
	val1 = func1(array1,array2,x)
	val2 = param/idX
	return val1*val2*array3*expX

def result(array1,array2,array3,param):
	def integrand(x):
		return func2(array1,array2,array3,param,x)
	Integral = scheme.integrate(integrand, [1.,10.])
	return Integral

res = result(array1,array2,array3,param)
```

Is this clearer ?

**Comment by nschloe** [comment]
Looks better than the above, but you can make it much shorter still. For example:
```python
import numpy as np
import quadpy

scheme = quadpy.c1.gauss_legendre(3)
array1 = np.ones((63, 101, 101))


def func2(x):
    return np.multiply.outer(x, array1)


scheme.integrate(func2, [1.0, 10.0])
```
gives
```
    assert fx.shape == range_shape + interval_set_shape + x.shape
AssertionError
```
_This_ is what I expect as the first message in an issue. Now the problem is totally clear, boiled down to a minimum. It's a bit of work to get there, for sure, but not much compared to me fixing a bug or adding a feature to quadpy.

**Comment by nschloe** [comment]
I had to look into it myself as well, so it's not really that clear. Solution: The dimensionality of the return values have to go first, so it has to be
```
    return np.multiply.outer(array1, x)
```
in the minimal example to work. See also https://github.com/nschloe/quadpy/wiki/Dimensionality-of-input-and-output-arrays.

**Comment by ACharletENSL** [comment]
It works now, thanks for your help and your patience
